source/common/upstream/upstream_impl.h

```diff
@@ -919,7 +919,8 @@ class BaseDynamicClusterImpl : public ClusterImplBase {
   bool updateDynamicHostList(const HostVector& new_hosts, HostVector& current_priority_hosts,
                              HostVector& hosts_added_to_current_priority,
                              HostVector& hosts_removed_from_current_priority,
-                             HostMap& updated_hosts, const HostMap& all_hosts);
+                             HostMap& updated_hosts, const HostMap& all_hosts,
+                             const absl::flat_hash_set<std::string>& all_new_hosts);
```

Comment:
Please add new doc comment on this above. I've been looking at this code recently and as you pointed out, it's already incredibly hard to understand. Any additional comments/corrections you can make would be very helpful.

Reply:
Also, as an aside, I'm always concerned about the perf impact to changes like this. It seems a shame we need yet another map/etc. One possible idea: Since this set is only used for temporary comparisons, could it be a flat_hash_set of string views to avoid copying? I think all of the string should live until the update function is done?

Reply:
Sorry did you look at this to see if we can use a hash_set of string_view vs. copied strings?

/wait-any

Reply:
I'm not sure I can use a string_view here, although it's new to me, so I may be misunderstanding it's usage. I'm not sure I can assume liveness of the underlying data (See e.g. eds.cc:63 or redis_cluster.cc:101 for how that set is populated.)

I've switched new_hosts_for_current_priority to use a string_view as that should be safe (upstream_impl.cc:1419), although I don't know if that improves anything?

Reply:
I had second thoughts about new_hosts_for_current_priority too so reverted changing that.

Reply:
OK yeah I see that we aren't using stable strings in all places, which is unfortunate. I think there is a separate effort to do perf analysis of some of the xDS paths so I guess we can look at this independently. cc @htuch @jmarantz @pgenera @rojkov